src/dmd/dmodule.d

```diff
@@ -536,6 +537,20 @@ extern (C++) final class Module : Package
         }
         m = m.parse();
 
+        // verify the module name matches the imported module name
+        if (m.md is null)
+        {
+            // in this case there was no module declaration, this means the import
+            // should have no package prefix, but for now we support this with a
+            // depecration message backwards compatibility
+            if (packages && packages.dim > 0)
+                m.deprecation(loc, "from file %s should be imported with 'import %s;'",
+                    m.srcfile.name.toChars(), m.toPrettyChars());
+        }
+        else if (!packages.opEquals(m.md.packages) || ident != m.md.id)
+            m.error(loc, "from file %s must be imported with 'import %s;'",
```

Comment:
This should probably go through a deprecation cycle.

Reply:
Maybe.  I think that some variations are more likely to be used out in the wild than others.  Here's my categorization of the various cases:

1. import a module with a qualified name and the module being imported has no module declaration

```D
// main.d
import foo.bar;
// foo/bar.d
// note the missing module declaration
```
This PR doesn't break this case but it does print a deprecation message.

2. import a module with a qualified name that partially matches the name of the module being imported

```D
// main.d
import foo1.bar;
import foo2.bar.baz.
// foo1/bar.d
module bar;
// foo2/baz/buz.d
module baz.buz;
```
Currently this PR makes these cases errors, however, I could see an argument to deprecate these instead.  doing so would be a simple modification to this PR, what do people think?

3. importing a module that matches the filename but does not match the module name
```D
// main.d
import foo;
// foo.d
module bar;
```
This is one case that I don't think we need to deprecate. I say kill it with fire and extreme prejudice.

Reply:
Note that this doesn't break the following case:
```
// main.d
import bar;
// foo.d
module bar;
```